### PR TITLE
[INFRA] apply black, flake8, pyupgrade to `nilearn/decoding`

### PR DESCRIPTION
**Pull request by Remi-Gau**
Relates to #2528

Changes proposed in this pull request:

- Set up pre-commit hooks to also work on the `nilearn/decoding` module
- Update config and GH worflows accordingly
- run `pre-commit run -a` and manually fix or silence errors
- convert some "old strings" to f-strings using flynt
- remove some lambdas (PEP8 E731)
- add some explicit line breaks in doc
- fixes some headings underline length in doc strings

### CODE REVIEW

**Comment by Remi-Gau** [comment]
List of potential refactoring suggested by sourcery: https://github.com/Remi-Gau/nilearn/pull/5/files

File: nilearn/decoding/tests/test_fista.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_fista.py:62]: probably need a better variable name

File: nilearn/decoding/tests/test_graph_net.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_graph_net.py:275]: probably need a better variable name

Comment (Remi-Gau) [reply nilearn/decoding/tests/test_graph_net.py:275]: avoid lambda

File: nilearn/decoding/tests/test_fista.py
Comment (Remi-Gau) [reply nilearn/decoding/tests/test_fista.py:62]: avoid lambda

File: nilearn/decoding/tests/test_fista.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_fista.py:58]: avoid lambda

File: nilearn/decoding/tests/test_fista.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_fista.py:59]: avoid lambda

File: nilearn/decoding/tests/test_graph_net.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_graph_net.py:138]: avoid lambda

File: nilearn/decoding/tests/test_graph_net.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_graph_net.py:142]: avoid lambda

File: nilearn/decoding/tests/test_graph_net.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_graph_net.py:164]: avoid lambda

File: nilearn/decoding/tests/test_graph_net.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_graph_net.py:161]: avoid lambda

**Review by Remi-Gau — COMMENTED** [review]
Just to highlight the flake8 errors

File: nilearn/decoding/tests/test_fista.py
Comment (bthirion) [reply nilearn/decoding/tests/test_fista.py:62]: Sorry, why should we avoid lambda here (and in the next lines) ?

Comment (Remi-Gau) [reply nilearn/decoding/tests/test_fista.py:62]: PEP8 E731

https://peps.python.org/pep-0008/#programming-recommendations

Always use a def statement instead of an assignment statement that binds a lambda expression directly to an identifier:

# Correct:
```
def f(x): return 2*x
```

# Wrong:
```
f = lambda x: 2*x
```

The first form means that the name of the resulting function object is specifically ‘f’ instead of the generic ‘<lambda>’. This is more useful for tracebacks and string representations in general. The use of the assignment statement eliminates the sole benefit a lambda expression can offer over an explicit def statement (i.e. that it can be embedded inside a larger expression)


see also:
https://stackoverflow.com/questions/25010167/e731-do-not-assign-a-lambda-expression-use-a-def

File: nilearn/decoding/space_net_solvers.py
Comment (bthirion) [inline nilearn/decoding/space_net_solvers.py:237]: small rant: this is not super readable

File: nilearn/decoding/space_net.py
Comment (bthirion) [inline nilearn/decoding/space_net.py:213]: small rant: this is not super readable

**Review by bthirion — COMMENTED** [review]
Impressive ! Thank you. 
I think it's basically OK for merge.

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:268]:
```suggestion
        """Debias w by rescaling the coefficients by a fixed factor.
```

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:695]: You need to explicitly break the line with a "\\" character, otherwise the rendering doc will only have the first line in bold and the second half will render in the description.

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:1156]: same comment here

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:1243]: same

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:1402]: same

File: nilearn/decoding/space_net.py
Comment (NicolasGensollen) [inline nilearn/decoding/space_net.py:1476]: same

**Comment by Remi-Gau** [comment]
> Impressive ! 

I am not good at taking compliments but I just want to highlight that I pretty much only had to add a `*` in a couple of config files and then run `pre-commit run -a` to let all of that happen.

File: nilearn/decoding/space_net_solvers.py
Comment (NicolasGensollen) [reply nilearn/decoding/space_net_solvers.py:326]: I agree, "l" is the worst possible variable name...

File: nilearn/decoding/space_net_solvers.py
Comment (Remi-Gau) [reply nilearn/decoding/space_net_solvers.py:237]: yeah that's the problem with black as a formatter, it is VERY opinionated.

I could add a `fmt: off` to those sections to tell black to leave it alone.

**Comment by Remi-Gau** [comment]
FYI: still need to fix the flake8 errors I have just silenced for now.

Leaving a bunch a `noqa` in the test code base feels like a [broken window](https://blog.codinghorror.com/the-broken-window-theory/) to me.

**Review by NicolasGensollen — COMMENTED** [review]
Great ! Thanks @Remi-Gau ! 👍 
LGTM minus the small issue for the rendered doc (see below).

File: nilearn/decoding/space_net.py
Comment (Remi-Gau) [reply nilearn/decoding/space_net.py:695]: Was not aware of that. Thanks.

File: nilearn/decoding/space_net_solvers.py
Comment (alexisthual) [reply nilearn/decoding/space_net_solvers.py:237]: It's just my opinion, but I wouldn't use `fmt: off` anywhere.
I think the whole purpose of using an automated formater is to never deal with formatting again. Indeed, there will be some ugly newlines, but overall the codebase will be much more readable, so I wouldn't care too much about these.

Comment (Remi-Gau) [reply nilearn/decoding/space_net_solvers.py:237]: yeah that's be my opinion too, but I wanted to make sure people are of other possibilities so we can make the best informed collective decision. 😉

File: nilearn/decoding/space_net_solvers.py
Comment (Remi-Gau) [reply nilearn/decoding/space_net_solvers.py:326]: any suggestion as to what to replace it with?

File: nilearn/decoding/space_net_solvers.py
Comment (alexisthual) [reply nilearn/decoding/space_net_solvers.py:237]: What we could do though is to increase the line length to the default line length in black and disable flake8's error. I guess this would significantly reduce the number of ugly lines.

File: nilearn/decoding/tests/test_fista.py
Comment (Remi-Gau) [inline nilearn/decoding/tests/test_fista.py:65]: are `*args` and `**kwargs` really needed here?

File: nilearn/decoding/space_net.py
Comment (Remi-Gau) [reply nilearn/decoding/space_net.py]: I fixed all those line breaks I had added.

Unrelated: when looking at the doc in html after building locally, it seems that the first line in bold is not a systematic thing for all arguments, no?

File: nilearn/decoding/space_net_solvers.py
Comment (ymzayek) [reply nilearn/decoding/space_net_solvers.py:237]: I don't have a strong opinion here. We can bring it up at the next developers meeting in case someone has a strong argument to keep it as is. I guess this will also be a good opportunity to discuss other formatting choices or changes to move forward with before reformatting the whole codebase

File: nilearn/decoding/tests/test_fista.py
Comment (alexisthual) [reply nilearn/decoding/tests/test_fista.py:65]: They might not be used in the function but be needed because some calls to this function specify kwargs which are not w or l, so I'd probably keep *args and **kwargs here.

File: nilearn/decoding/space_net_solvers.py
Comment (ymzayek) [reply nilearn/decoding/space_net_solvers.py:326]: Maybe `stepsize`? I see that here in the definition of f2_prox: https://github.com/nilearn/nilearn/blob/b1fa2e0b9161a5055a900b708aed38aaab15866f/nilearn/decoding/fista.py#L79

Comment (Remi-Gau) [reply nilearn/decoding/space_net_solvers.py]: done

**Comment by Remi-Gau** [comment]
I think I am pretty much done with this PR. Unless I missed something or some comments.

**Review by NicolasGensollen — APPROVED** [review]
LGTM, thanks @Remi-Gau !

**Review by ymzayek — COMMENTED** [review]
LGTM also! I just wonder if we want to reconsider line length before merging. We can discuss it tomorrow at the coredev meeting.

**Comment by Remi-Gau** [comment]
OK another "minor" point that may require some discussions:
- where to put the default value for parameters in doc strings.

In this PR, I have started to move the default value I found sometimes at the end of the "description" section of a parameter doc string, to the first line of the doc string (with the type declaration).

This seems to be what is shown the numpy doc style.
https://numpydoc.readthedocs.io/en/latest/format.html#parameters

It would also be more convenient if at any point in the future we want to add / enforce proper type annotation.

Example of one of the change I made (bold is just for the example)

```rst
mask: filename, Nifti1Image, NiftiMasker, or MultiNiftiMasker, optional,\
    Mask to be used on data. If an instance of masker is passed,
    then its mask and parameters will be used. If no mask is given, mask
    will be computed automatically from provided images by an inbuilt
    masker with default parameters. Refer to NiftiMasker or
    MultiNiftiMasker to check for default parameters. **Default None**
```

```rst
mask: filename, Nifti1Image, NiftiMasker, or MultiNiftiMasker, optional,\
    **(Default None)**
    Mask to be used on data. If an instance of masker is passed,
    then its mask and parameters will be used. If no mask is given, mask
    will be computed automatically from provided images by an inbuilt
    masker with default parameters. Refer to NiftiMasker or
    MultiNiftiMasker to check for default parameters.
```

If a consistent style is chosen, it may be worth adding opening an issues and adding that to the list of things to do in the "fade to black / spring cleaning".

**Comment by bthirion** [comment]
We need to to the rebase after merging of  #3508.

**Review by bthirion — APPROVED** [review]

**Comment by ymzayek** [comment]
@Remi-Gau we discussed line length and placement of the default value yesterday at the coredev meeting. 

For line length it was agreed to keep it as is. There was the point that increasing line length might not necessarily lead to black making better choices in terms of readability and not many lines are affected in this way anyways. 

Concerning the default value we agreed to follow the numpydoc style guide. We were also curious about your type annotation comment as it was understood in different ways (one relating to type hinting and the other to consistency between the docstring default value and function signature). I think it would be nice to discuss with you directly (perhaps at one of the [drop-in hours](https://arewemeetingyet.com/UTC/2023-01-18/16:00/w/Nilearn%20Drop-in%20Hours#eyJ1cmwiOiJodHRwczovL2Rpc2NvcmQuZ2cvYk1CaGI3dyJ9) soon).

Also curious about your opinion about these choices.

**Comment by Remi-Gau** [comment]
> @Remi-Gau we discussed line length and placement of the default value yesterday at the coredev meeting.
> 
> For line length it was agreed to keep it as is. There was the point that increasing line length might not necessarily lead to black making better choices in terms of readability and not many lines are affected in this way anyways.

Good with me. 
Then I think this is PR is done, no?

> Concerning the default value we agreed to follow the numpydoc style guide. We were also curious about your type annotation comment as it was understood in different ways (one relating to type hinting and the other to consistency between the docstring default value and function signature). I think it would be nice to discuss with you directly (perhaps at one of the [drop-in hours](https://arewemeetingyet.com/UTC/2023-01-18/16:00/w/Nilearn%20Drop-in%20Hours#eyJ1cmwiOiJodHRwczovL2Rpc2NvcmQuZ2cvYk1CaGI3dyJ9) soon).
> 
> Also curious about your opinion about these choices.

FYI

Regarding where to put the default value.
Had a small chat with @jeromedockes about this.
He rightly suggested another possibility: only keep the default value mentioned in the signature of the function which avoids any eventual inconsistency between the signature and the docstring.

We briefly talked about adding type annotation: in brief filed that mostly under "nice to have in the long run" and would have to be done carefully with considerations to potential API breakage, raising the technical bar for new contributors, unpractical if many dependencies do not have type annotations...

**Comment by ymzayek** [comment]
> He rightly suggested another possibility: only keep the default value mentioned in the signature of the function which avoids any eventual inconsistency between the signature and the docstring.

Ok, will this be handled in another PR? If so, can you open an issue so we can track it.

Otherwise I think yes, this one should be good. Thx @Remi-Gau !

**Comment by Remi-Gau** [comment]
> Ok, will this be handled in another PR? 
 
I prefer to handle that in a different PRs: those black + flake8 PRs are already big enough as it is, so I would prefer to put most of the automated changes in separated PRs from the stylistic changes that require manual edits

> If so, can you open an issue so we can track it.

Will do.

**Comment by ymzayek** [comment]
Great merging then

**Comment by bthirion** [comment]
Thx !

**Pull request merged**
